infra/rooch-portal-v2/contract/gas_market/sources/gas_market.move

```diff
@@ -0,0 +1,166 @@
+    }
+
+    // 0.1 U = BTC decimal / (RGas decimal * 10)
+    const DEFAULT_UNIT_PRICE: u256 = 1;
+    const BTC_USD: vector<u8> = b"BTCUSD";
+
+    const ErrorMarketNotOpen: u64 = 0;
+    const ErrorReceiverAddress: u64 = 1;
+    const ErrorTokenPrice: u64 = 2;
+    const ErrorNoUncheckTxid: u64 = 3;
+
+
+    fun init() {
+        let rgas_market_obj = object::new_named_object(RGasMarket{
+            rgas_store: coin_store::create_coin_store<RGas>(),
+            unit_price: DEFAULT_UNIT_PRICE,
+            receive_btc_address: sender(),
+            is_open: true
+        });
+        let admin_cap = object::new_named_object(AdminCap{});
+
+        move_resource_to(&module_signer<MarketInfo>(), MarketInfo{
+            total_deposit: 0,
+            total_withdraw: 0,
+            buyer: table::new(),
+            uncheck_info: table::new()
+        });
+
+        to_shared(event_queue::subscribe<ReceiveUTXOEvent>(to_string(&sender())));
+        transfer(admin_cap, sender());
+        to_shared(rgas_market_obj)
+    }
+
+    public entry fun add_rgas_coin(
+        account: &signer,
+        rgas_market_obj: &mut Object<RGasMarket>,
+        amount: u256
+    ){
+        let rgas_market = object::borrow_mut(rgas_market_obj);
+        assert!(rgas_market.is_open, ErrorMarketNotOpen);
+        let rgas_coin = account_coin_store::withdraw<RGas>(account, amount);
+        coin_store::deposit(&mut rgas_market.rgas_store, rgas_coin);
+        let market_info_mut = borrow_mut_resource<MarketInfo>(address_of(&module_signer<MarketInfo>()));
+        market_info_mut.total_deposit = market_info_mut.total_deposit + amount
+    }
+
+    public entry fun withdraw_rgas_coin(
+        _admin: &mut Object<AdminCap>,
+        rgas_market_obj: &mut Object<RGasMarket>,
+        amount: u256
+    ){
+        let rgas_market = object::borrow_mut(rgas_market_obj);
+        assert!(rgas_market.is_open, ErrorMarketNotOpen);
+
+        let rgas_coin = coin_store::withdraw<RGas>(&mut rgas_market.rgas_store, amount);
+        account_coin_store::deposit(sender(), rgas_coin);
+        let market_info_mut = borrow_mut_resource<MarketInfo>(address_of(&module_signer<MarketInfo>()));
+        market_info_mut.total_withdraw = market_info_mut.total_withdraw + amount
+    }
+
+    public entry fun consume_event(
+        rgas_market_obj: &mut Object<RGasMarket>,
+        subscriber_obj: &mut Object<Subscriber<ReceiveUTXOEvent>>
+    ){
+        let rgas_market = object::borrow_mut(rgas_market_obj);
+        assert!(rgas_market.is_open, ErrorMarketNotOpen);
+        let consume_event = option::extract(&mut consume(subscriber_obj));
+        let (txid, sender, receiver, value) = unpack_receive_utxo_event(consume_event);
+        assert!(receiver == rgas_market.receive_btc_address, ErrorReceiverAddress);
+        let price_info = trusted_price(utf8(BTC_USD));
+        let token_price = value(&price_info);
+        // BTC decimal is 10**9 RGas decimal is 10**8 RGas price is 0.1U
```

Comment:
BTC decimal also is 8, RGas price is 0.01U, 100 RGas = 1U

Reply:
Provide a read-only function to calculate the buy result, two functions:

btc_to_rgas(BTC):RGas
rgas_to_btc(RGas):BTC